python/lsst/pipe/base/graph/_loadHelpers.py

```diff
@@ -0,0 +1,310 @@
+from typing import (Optional, Iterable, DefaultDict, Set, Dict, TYPE_CHECKING, Type, Union)
+
+if TYPE_CHECKING:
+    from . import QuantumGraph
+
+
+# Create a custom dict that will return the desired default if a key is missing
+class RegistryDict(UserDict):
+    def __missing__(self, key):
+        return DefaultLoadHelper
+
+
+# Create a registry to hold all the load Helper classes
+HELPER_REGISTRY = RegistryDict()
+
+
+def register_helper(URICLass: Union[Type[ButlerURI], Type[io.IO[bytes]]]):
+    """Used to register classes as Load helpers
+
+    When decorating a class the parameter is the class of "handle type", i.e.
+    a ButlerURI type or open file handle that will be used to do the loading.
+    This is then associated with the decorated class such that when the
+    parameter type is used to load data, the appropriate helper to work with
+    that data type can be returned.
+
+    A decorator is used so that in theory someone could define another handler
+    in a different module and register it for use.
+
+    Parameters
+    ----------
+    URIClass : Type of `~lsst.daf.butler.ButlerURI` or `~io.IO` of bytes
+        type for which the decorated class should be mapped to
+    """
+    def wrapper(class_):
+        HELPER_REGISTRY[URICLass] = class_
+        return class_
+    return wrapper
+
+
+class DefaultLoadHelper:
+    """Default load helper for `QuantumGraph` save files
+
+    This class, and its subclasses, are used to unpack a quantum graph save
+    file. This file is a binary representation of the graph in a format that
+    allows individual nodes to be loaded without needing to load the entire
+    file.
+
+    This default implementation has the interface to load select nodes
+    from disk, but actually always loads the entire save file and simply
+    returns what nodes (or all) are requested. This is intended to serve for
+    all cases where there is a read method on the input parameter, but it is
+    unknown how to read select bytes of the stream. It is the responsibility of
+    sub classes to implement the method responsible for loading individual
+    bytes from the stream.
+
+    Parameters
+    ----------
+    uriObject : `~lsst.daf.butler.ButlerURI` or `io.IO` of bytes
+        This is the object that will be used to retrieve the raw bytes of the
+        save.
+    """
+    def __init__(self, uriObject: Union[ButlerURI, io.IO[bytes]]):
+        self.uriObject = uriObject
+
+        # Read the first 18 bytes which correspond to the 2 byte version
+        # number and the two 8 bytes numbers that are the sizes of the byte
+        # maps
+        sizeBytes = self.uriObject.read(18)
+
+        # Turn they encode bytes back into a python int object
+        save_version, taskDefSize, nodeSize = struct.unpack('>HQQ', sizeBytes)
+
+        # Store the save version, so future read codes can make use of any
+        # format changes to the save protocol
+        self.save_version = save_version
```

Comment:
Im not sure what you mean about "magic string", and which other comment are you referring to?

Reply:
After changing the code anyway I think I understand what you meant by this, and well its changed to stay read the string from the graph module now.

Reply:
"magic" can be any string that can help you uniquely identify file format. Right now you expect that file header is just a three numbers, but if by mistake someone gives you a random file that starts with a sequence `\0x00\0x01` then you will still try to interpret as a valid file. It will fail in pickle somewhere but it may be nicer to be able to identify that file format is wrong by choosing relatively unique bytes for header, e.g. `"qgraph\0x00\0x01...."`